all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+`NumWorkflowTaskPollers` (or routines), what it probably should be is `MaxConcurrentWorkflowTaskPolls`, since the number
+of open RPC calls is actually what's fluctuating.
+
+
+### Slot tuning & Cache Size
+
+Workers have a configurable number of slots for task types. In Core: `max_outstanding_workflow_tasks`,
+`max_outstanding_activities`, and `max_outstanding_local_activities`. These are set by the user and can't change at
+runtime. Autotuning of these values seeks to allow them to change at runtime according to available resources on
+the worker. It is not feasible for us to choose *optimal* values for these at runtime, because we cannot know a-priori
+about how much resources any given workflow or activity task will consume - only the user can (maybe) know this.
+However, we can provide reasonable default autotuning implementations while also allowing users to provide their own.
+
+Additionally, there is a maximum cache size for cached workflows. In Core: `max_cached_workflows`. This is also set by
+users at the moment, and autotuning it is subject to the same lack of knowledge we have as with slot tuning.
+
+## Constraints
+
+There are some invariants that must be maintained:
+
+* The current number of active polls for a task type must be <= the number of free slots for that task type at all
+  times. Otherwise, we can end up receiving a task that won't have anywhere to go until a slot frees. For example,
+  in Core this is expressed by using a permit reservation system.
+* Max outstanding workflow tasks must be <= max cached workflows.
+
+## Proposed interface for slot management
+
+To reiterate, we can't possibly know as well as the user what resources a given workflow or activity task will consume.
+Thus, we need to provide an interface by which users can tell us whether they think it's reasonable to allow the use
+of an additional slot.
+
+The interface definition here is provided in Rust, but need not be copied exactly by other languages. Since users are
+implementing it, we want something idiomatic for them. Rust will likely immediately convert all newly allowed slots into
+a permit struct which will automatically release the slot when dropped (like it does today, without this interface). 
+Languages with destructor semantics should do the same, as permit/slot leaking is easy to do by accident.
+
+There is some mild generic magic going on to represent different types of slots and constrain that the right info type
+is provided. These guarantees can't be compile-time enforced across the language boundary, but are at least useful after
+runtime validation.
+
+```rust
+trait SlotSupplier {
+  type SlotKind: SlotKind;
+  /// Blocks until a slot is available.
+  /// May also return an error if the backing user implementation encounters an error.
+  /// Returned errors will bubble up and cause the worker to shut down.
+  async fn reserve_slot(&self) -> Result<(), anyhow::Error>;
```

Comment:
I think `reserve_slot` should take some sort of cancellation token (what is very lang dependent) and also return a bool if it did not acquire a slot. `reserve_slot` needs to be cancellable for the worker shutdown case, if the number of pollers is scaling down due to poller autotuning, and the whole eager workflow/activity use case. It should return a bool as well since in the later two cases a slot was not reserved, but I do not want the whole worker to shutdown

Reply:
>I don't totally follow how the number of pollers scaling down works with that though

So if the number of pollers is scaling down, a `reserve_slot` call may be cancelled and not reserve a slot, but also not error (since that would kill the worker)

Reply:
Right, but you mean it's cancelled by the caller, right? That explains some of the confusion here since in Rust you'd just drop the future and therefore never see the result.

That's why I didn't have a bool. If the _implementer_ would return false, IMO they should just keep blocking. So, even in Go/.NET with a stop channel or cancel token, we could also have this return void since in those situations if the caller cancels the call before it's finished, they should also just ignore the result.

Reply:
Ignoring the result is a race condition since it is possible to receive a slot even if tried to cancel the call. You need to know if it suceeded/failed

Reply:
I was thinking that, but, I figured since you're waiting on that call to resolve you would either have seen it resolve before you issued the cancel or not, and if it wasn't resolved before that means to the implementer that is *must* not count as having been reserved

But... I can see how that is difficult to guarantee all the time.

So anyway, tldr on all this yes I agree it needs to be like that in those langs. In Rust it can just return `()` and the future can be dropped if the reservation is no longer desired and that's easy to do race-free. I'll add comments to that effect.